tests/commands/test_build.py

```diff
@@ -0,0 +1,692 @@
+    """It's logged in error if cmd fails."""
+    caplog.set_level(logging.ERROR, logger="charmcraft")
+
+    cmd = ['python3', '-c', "exit(3)"]
+    retcode = polite_exec(cmd)
+    assert retcode == 3
+    assert any("Execution ended in 3" in rec.message for rec in caplog.records)
+
+
+def test_politeexec_crashed(caplog, tmp_path):
+    """It's logged in error if cmd fails."""
+    caplog.set_level(logging.ERROR, logger="charmcraft")
+    nonexistent = tmp_path / 'whatever'
+
+    cmd = [str(nonexistent)]
+    retcode = polite_exec(cmd)
+    assert retcode == -1
+    assert any("Execution crashed with FileNotFoundError" in rec.message for rec in caplog.records)
+
+
+# --- (real) build tests
+
+
+def test_build_basic_complete_structure(tmp_path):
+    """Integration test: a simple structure with custom lib and normal src dir."""
+    build_dir = tmp_path / BUILD_DIRNAME
+    build_dir.mkdir()
+
+    # the metadata
+    metadata = tmp_path / 'metadata.yaml'
+    with metadata.open('wb') as fh:
+        fh.write(b'lot of yaml config')
+
+    # a lib dir
+    lib_dir = tmp_path / 'lib'
+    lib_dir.mkdir()
+    ops_lib_dir = lib_dir / 'ops'
+    ops_lib_dir.mkdir()
+    ops_stuff = ops_lib_dir / 'stuff.txt'
+    with ops_stuff.open('wb') as fh:
+        fh.write(b'ops stuff')
+
+    # simple source code
+    src_dir = tmp_path / 'src'
+    src_dir.mkdir()
+    charm_script = src_dir / 'charm.py'
+    with charm_script.open('wb') as fh:
+        fh.write(b'all the magic')
+
+    builder = Builder({
+        'from': pathlib.Path(str(tmp_path)),  # bad support for tmp_path's pathlib2 in Py3.5
+        'entrypoint': charm_script,
+        'requirement': [],
+    })
+    zipname = builder.run()
+
+    # check all is properly inside the zip
+    # contents!), and all relative to build dir
+    zf = zipfile.ZipFile(zipname)
+    assert zf.read('metadata.yaml') == b"lot of yaml config"
+    assert zf.read('src/charm.py') == b"all the magic"
+    dispatch = DISPATCH_CONTENT.format(entrypoint_relative_path='src/charm.py').encode('ascii')
+    assert zf.read('dispatch') == dispatch
+    assert zf.read('hooks/install') == dispatch
+    assert zf.read('hooks/start') == dispatch
+    assert zf.read('hooks/upgrade-charm') == dispatch
+    assert zf.read('lib/ops/stuff.txt') == b"ops stuff"
+
+
+def test_build_code_simple(tmp_path):
+    """Check transferred metadata and simple entrypoint."""
+    build_dir = tmp_path / BUILD_DIRNAME
+    build_dir.mkdir()
+
+    metadata = tmp_path / CHARM_METADATA
+    entrypoint = tmp_path / 'crazycharm.py'
+
+    builder = Builder({
+        'from': tmp_path,
+        'entrypoint': entrypoint,
+        'requirement': [],
+    })
+    linked_entrypoint = builder.handle_code()
+
+    built_metadata = build_dir / CHARM_METADATA
+    assert built_metadata.is_symlink()
+    assert built_metadata.resolve() == metadata
+
+    built_entrypoint = build_dir / 'src' / 'crazycharm.py'  # note it's inside 'src'
```

Comment:
as mentioned before, I don't prefer this.

Reply:
Already changed, yes, thanks